### PR TITLE
Implement constraints for nn.Parameters of nn.Modules

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2078 

This implements a surprisingly succinct syntax for constraints of module parameters, e.g.
```py
from torch import nn
from torch.distributions import constraints
from pyro.params import constraint

class Normal(nn.Module):
    def __init__(self, loc, scale):
        super().__init__()
        self.loc = loc                 # this will be unconstrained
        self.scale = scale             # this will be constrained

    @constraint
    def scale(self):                   # must match name of attribute
        return constraints.positive    # a constraint object,
                                       # can even depend on other params
```
Hopefully this moves us towards making the Pyro param store a nn.Module. Currently this is not used by the ParamStore, but this PR will unblock later refactoring.

In the nearer term, @martinjankowiak might want to use this for constrained parameters in pyro.contrib.timeseries, e.g. this could simplify #2099.

## How does this work?

For each constrained parameter `x`, this creates an unconstrained parameter with name `x_unconstrained`, which is accessible in the usual way. The name `x` then points to a Python descriptor (like a property) that transforms to/from `unconstrained_x` to a user-defined value. Thus all the usual `nn.Module` magic should continue to work correctly, with only the addition of this `transform_to()` layer.

## Tested
- [x] tested with a variety of constraints
- [x] tested chaining, where one param's constraint might depend on another param (e.g. for lower bound)
- [x] tested serialization

### CODE REVIEW

**Comment by martinjankowiak** [comment]
to clarify, if such an object is an attribute of some `module` this will support methods like `module.cuda()` right?

**Comment by fritzo** [comment]
Yes, this will support methods like `module.cuda()`.

File: pyro/params/constrained_parameter.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
+import torch
+from torch.distributions import transform_to
+
+
+class ConstrainedParameter:
+    """
+    Descriptor to add a constraint to a
```
Comment (neerajprad) [inline pyro/params/constrained_parameter.py]: nit: additional `to a` above.

**Comment by fritzo** [comment]
> will this require us to start storing parameters as attributes of ParamStore class?

Good question. I think that, independently of this PR, if we make the ParamStore an nn.Module, then yes we would need to store pyro parameters as attributes of the ParamStore. While that would be easy to do for unconstrained parameters, I do not know how to dynamically set constraints (this PR only allows static constraints). We should think more on this.

**Pull request merged**